// reactjs-bootstrap-table/lib/Resizer.js
'use strict';

module.exports = function (tableComponent, resizeObj, columnWidths) {

  var initialLoad = true;

  function debug(msg) {
    //console.log('BootstrapTable: ' + msg); 
  }

  function getSizeOfElements(ids) {
    if (!(ids && ids.length)) return 0;
    var total = 0;
    ids.forEach(function (id) {
      var elem = document.getElementById(id);
      if (elem && elem.offsetHeight) {
        total += elem.offsetHeight;
      }
    });
    return total;
  }

  function setHeaderWidths() {
    var widths = [];
    var tbody = document.getElementById(tableComponent.bodyId);
    var thead = document.getElementById(tableComponent.headerId);
    if (!(tbody && thead)) return;
    if (!(thead.rows.length && tbody.rows.length)) return;

    var cellWidths = [];
    for (var i = 0; i < tbody.rows[0].cells.length; i++) {
      var width = tbody.rows[0].cells[i].offsetWidth;
      cellWidths.push(width);
    }
    for (var _i = 0; _i < thead.rows[0].cells.length; _i++) {
      var _width = cellWidths[_i];
      var th = thead.rows[0].cells[_i];
      th.style.width = '' + _width + 'px';
    }
  }

  var id = tableComponent.id;
  var extra = 0;
  var minSize = 0;

  if (typeof resizeObj.minSize === 'number') {
    minSize = resizeObj.minSize;
  }

  if (typeof resizeObj.extra === 'number') {
    extra = resizeObj.extra;
  }
  if (resizeObj.elements && resizeObj.elements.length) {
    extra = extra + getSizeOfElements(resizeObj.elements);
  }
  if (extra) {
    var headerElement = document.getElementById(tableComponent.headerId);
    if (headerElement) {
      extra = extra + headerElement.offsetHeight;
    }
  }

  function resize() {
    var h = window.innerHeight;
    var table = document.getElementById(tableComponent.id);
    var tableBody = document.getElementById(tableComponent.bodyId);
    var tableHeader = document.getElementById(tableComponent.headerId);

    debug('Resize -> Window HEIGHT: ' + h);

    if (!table) {
      return;
    }

    var th = table.offsetHeight;
    var thh = tableHeader.offsetHeight;
    var tbh = tableBody.offsetHeight;

    var tw = table.offsetWidth;

    debug('Resize -> Table  offsetHeight: ' + th);
    debug('Resize -> Header offsetHeight: ' + thh);
    debug('Resize -> Body   offsetHeight: ' + tbh);

    if (extra) {
      th = h - extra;
    }
    if (th < minSize) {
      th = minSize;
    }

    if (!initialLoad) {
      debug("INITAL LOAD adjustment");
      initialLoad = false;
      th = th + thh;
    }
    tableBody.style.height = '' + th + 'px';
    debug('Resize -> Set Table Height to ' + th);
    setHeaderWidths();
  }

  this.addHandler = function () {
    window.addEventListener('resize', resize);
    resize();
  };

  this.removeHandler = function () {
    document.body.removeEventListener('resize', resize);
  };
};
